### PR TITLE
switch continuous integration system to Drone CI

### LINKED ISSUE #1649: Changes in Travis CI pricing
Hi Everyone,

I've just learnt that Travis CI is making changes to their pricing model that affects OSS projects, too. You might want to look into this depending on your usage stats: https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing

The original article that linked to the above post in the first paragraph: https://blog.healthchecks.io/2020/11/using-github-actions-to-run-django-tests/

> I recently found out Travis CI is ending its free-for-opensource offering, and looked at the alternatives.

### PR DESCRIPTION
**Pull request by nodiscc**
See commit messages

### CODE REVIEW

File: .drone.yml [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+# disable drone built-in clone action (always runs as root)
+clone:
+  disable: true
+steps:
+  - name: test-php-8.0
+    image: shaarli/drone-ci-base:8.0
+    # pull: always # force update of base image even if cached locally
+    commands:
+      # git clone as drone user
+      - git clone --depth=50 --branch=$DRONE_BRANCH $DRONE_GIT_HTTP_URL .
```
Comment (nodiscc) [inline .drone.yml:16]: `$DRONE_GIT_HTTP_URL` points to https://github.com/shaarli/Shaarli but the HEAD commit is only present on https://github.com/nodiscc/Shaarli

Comment (ArthurHoaro) [reply .drone.yml:16]: Wouldn't it fail on any PR, where the commit ID points to the submitter's fork and not Shaarli repository itself?

File: .drone.yml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+clone:
+  disable: true
+steps:
+  - name: test-php-7.4
+    image: shaarli/drone-ci-base:7.4
+    # pull: always # force update of base image even if cached locally
+    commands:
+      - git clone --depth=50 --branch=$DRONE_BRANCH $DRONE_GIT_HTTP_URL .
+      - git checkout $DRONE_COMMIT
+      - composer config --unset platform && composer config platform.php 7.4
```
Comment (ArthurHoaro) [inline .drone.yml:44]: Good!
Fixing the PHP version as we do isn't a very good practice. Once we've fixed the PHP 8.0 issue, we can probably remove that setting entirely from `composer.json`.

File: .drone.yml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+type: docker
+name: php-8.0
+
+# disable drone built-in clone action (always runs as root)
+clone:
+  disable: true
+steps:
+  - name: test-php-8.0
+    image: shaarli/drone-ci-base:8.0
+    # pull: always # force update of base image even if cached locally
```
Comment (ArthurHoaro) [inline .drone.yml:13]: Why is this commented? If it's not required we should remove it.

**Review by ArthurHoaro — COMMENTED** [review]
Thank you for work on that part! That's great!

Two other things:

  * could you open a separate improvement issue to use the matrix thingy of Drone, copy/pasting config usually leads to out of sync settings issues.
  * how is this doing with Composer/npm cache?

File: .drone.yml (hunk 1 above)
Comment (nodiscc) [reply .drone.yml:16]: yes

File: .drone.yml (hunk 3 above)
Comment (nodiscc) [reply .drone.yml:13]: ok. something to be aware of though, in case Dockerfiles at https://github.com/shaarli/drone-ci-base are updated

**Comment by ArthurHoaro** [comment]
@nodiscc Would you like some help on this? We are almost out of Travis credits.

**Comment by nodiscc** [comment]
> how is this doing with Composer/npm cache?

I don't know. Barebones drone doesn't provide cache as far I know. But there is obviously some sort of caching given the build times/logs I am getting - but I didn't configure anything.

At some point I tried to implement [volume cache](http://plugins.drone.io/drillster/drone-volume-cache/) and a `restore_cache` step, but all I was getting was cache misses. I think the CI instance/runner has to explicitly enable such plugins/volume mounting, and I'm pretty sure cloud.drone.io doesn't. Another solution is to use a cache plugin backed by S3 storage (paid, expensive).

> separate improvement issue to use the matrix thingy of Drone

jsonnet https://docs.drone.io/pipeline/kubernetes/examples/language/rust/#test-multiple-versions ? will do

> We are almost out of Travis credits.

:scream: already? damn sorry other things kept me busy. The only remaining problem is `$DRONE_GIT_HTTP_URL` always pointing to the base fork. It's caused by the workaround to prevent drone from cloning the repo as root (`clone:  disable: true` + manually implemented `git clone`). I'll look into it now but any pointers are welcome

**Comment by nodiscc** [comment]
Done, rebased. I removed the workaround for `git clone` always being run as root, it appears to work correctly/clone as non-root since `USER drone` was added to https://github.com/shaarli/drone-ci-base. Is there anything else that we should check?

**Comment by ArthurHoaro** [comment]
> I don't know. Barebones drone doesn't provide cache as far I know. But there is obviously some sort of caching given the build times/logs I am getting - but I didn't configure anything.

You are right, they probably have something internally, but in any case dependencies + all tests in 18 seconds is not bad at all.

> jsonnet docs.drone.io/pipeline/kubernetes/examples/language/rust/#test-multiple-versions ? will do

Sure it looks great.

> 😱 already? damn sorry other things kept me busy.

No worries, I didn't want to pressure you or anything. 

> Done, rebased. I removed the workaround for git clone always being run as root, it appears to work correctly/clone as non-root since USER drone was added to shaarli/drone-ci-base. Is there anything else that we should check?

Awesome work! 🙌 Thank you.

Everything looks good to me. I think that we can merge this now, and if there are other improvements to add, we can do so later.

**Review by ArthurHoaro — APPROVED** [review]

**Pull request merged**